# PanK-Pipeline: A Pan-genome _K_-mer Pipeline for Population Analysis

## Introduction

## Installation
The pipeline [PanK-Pipeline](https://github.com/caixu0518/PanK-Pipeline) is installation-free but requires dependencies: 

Required:
1. [jellyfish](https://github.com/gmarcais/Jellyfish). In the present pipeline,[jellyfish](https://github.com/gmarcais/Jellyfish) is mainly used to quickly generate _k_-mers from resequencing reads and perform _k_-mer query.
2. [bedtools](https://github.com/arq5x/bedtools2)(v2.27.1).

Optional：
1. [plink](https://www.cog-genomics.org/plink2/) (v1.90b6.21). [plink](https://www.cog-genomics.org/plink2/) is mainly used when performing population structure and PCA analysis with _k_-mers.
2. [VCF2Dis](https://doi.org/10.1093/gigascience/giaf032)(VCF2Dis-1.54). [VCF2Dis](https://doi.org/10.1093/gigascience/giaf032) is used to make the phylogenetic tree based on _k_-mer presence and absence matrix.
3. [faststructure](https://github.com/rajanil/fastStructure). [faststructure](https://github.com/rajanil/fastStructure) is used to make the population structure analysis. The present pipeline recomeneded a docker repository (dockerbiotools/faststructure).   

## Inputs

## Outputs


## PanK-Pipeline
Here, We applied the _Brassica rapa_ pangenome, comprising 30 assemblies, together with resequencing data from 1,543 accessions, to demonstrate the [PanK-Pipeline](https://github.com/caixu0518/PanK-Pipeline) pipeline. This pipeline is broadly applicable to pangenome assemblies and population resequencing data from any species.

### Step1. Pipeline for constructing polymorphic _k_-mers using 30 _B. rapa_ genome assemblies.
The current script Generate_PolymorphyicKmers.pl is used for constructing polymorphic _k_-mers using Pan-genome assemblies
```
perl  Generate_PolymorphyicKmers.pl  -species Brapa  -ksize 17  -pangenome  Pangenome.txt  -PipelinePath  /mydata/caix/PanK-Pipeline

-species        [required]    Prefix for output files    i.e. Brapa
-ksize          [required]    k-mer size  i.e 17
-pangenome      [required]    A file containing the abbreviation of each pan-genome member together with the absolute path to its genome sequence.
-PipelinePath   [required]    The absolute path to the PanK-Pipeline directory   i.e. /mydata/caix/PanK-Pipeline

##- A detailed example of the '-pangenome' parameter input file: the first column specifies the abbreviation, and the second column specifies the absolute path to the genome FASTA file, with columns separated by a tab.
Bra_A03.A04	/mydata/caix/PanK-Pipeline/data/aa/new/Bra_A03.A04.fasta
Bra_BRO.A04	/mydata/caix/PanK-Pipeline/data/aa/new/Bra_BRO.A04.fasta
Bra_CCA.A04	/mydata/caix/PanK-Pipeline/data/aa/new/Bra_CCA.A04.fasta
```

### Step2: Pipeline for identifying representative k-mers across the _B. rapa_ species.
```
perl  Generate_RepresentativeKmers.pl  -species  Brapa  -ksize  17  -pangenome  Pangenome.txt    -PipelinePath   /mydata/caix/PanK-Pipeline   -PolymorphicKmer  rapa.merged.kmer.k17.list.Polymorphic_kmers.List
-species            [required]    Prefix for output files    i.e. Brapa
-ksize              [required]    k-mer size  i.e 17
-pangenome          [required]    A file containing the abbreviation of each pan-genome member together with the absolute path to its genome sequence.
-PipelinePath       [required]    The absolute path to the PanK-Pipeline directory   i.e. /mydata/caix/PanK-Pipeline
-PolymorphicKmer    [required]    rapa.merged.kmer.k17.list.Polymorphic_kmers.List

##- _rapa.merged.kmer.k17.list.Polymorphic_kmers.List_ was generated by the step 1.

```
### Step3：Application of species-representative _k_-mers for population structure analysis in _B. rapa_.
```

```


## Citations


